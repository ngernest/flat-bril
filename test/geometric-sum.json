{
  "functions": [
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "r",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "r_n",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "i",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "label": "loop"
        },
        {
          "args": [
            "i",
            "n"
          ],
          "dest": "cond",
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "cond"
          ],
          "labels": [
            "body",
            "done"
          ],
          "op": "br"
        },
        {
          "label": "body"
        },
        {
          "args": [
            "r_n",
            "r"
          ],
          "dest": "r_n",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "i",
            "one"
          ],
          "dest": "i",
          "op": "add",
          "type": "int"
        },
        {
          "labels": [
            "loop"
          ],
          "op": "jmp"
        },
        {
          "label": "done"
        },
        {
          "args": [
            "one",
            "r_n"
          ],
          "dest": "one_minus_r_n",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "a",
            "one_minus_r_n"
          ],
          "dest": "numerator",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "one",
            "r"
          ],
          "dest": "one_minus_r",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "numerator",
            "one_minus_r"
          ],
          "dest": "sum",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "sum"
          ],
          "op": "print"
        }
      ],
      "name": "main"
    }
  ]
}
