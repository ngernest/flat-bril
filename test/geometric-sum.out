@main
	op: const		dest: one		type: int		value: 1	
	op: const		dest: r_n		type: int		value: 1	
	op: const		dest: i 		type: int		value: 0	
.loop
	op: lt   		dest: cond		type: bool		args: ["i", "n"]	
	op: br   		args: ["cond"]		labels: ["body", "done"]
.body
	op: mul  		dest: r_n		type: int		args: ["r_n", "r"]	
	op: add  		dest: i 		type: int		args: ["i", "one"]	
	op: jmp  		labels: ["loop"]
.done
	op: sub  		dest: one_minus_r_n		type: int		args: ["one", "r_n"]	
	op: mul  		dest: numerator		type: int		args: ["a", "one_minus_r_n"]	
	op: sub  		dest: one_minus_r		type: int		args: ["one", "r"]	
	op: div  		dest: sum		type: int		args: ["numerator", "one_minus_r"]	
	op: print		args: ["sum"]	
